getComputedStyle(img).height returns string of a rounded int not a float

sammygill opened the pull request:
#### e2c4f937d459a12db2d03ff534c19f20828a6ecc
<pre>
getComputedStyle(img).height returns string of a rounded int not a float
<a href="https://bugs.webkit.org/show_bug.cgi?id=206161">https://bugs.webkit.org/show_bug.cgi?id=206161</a>
&lt;rdar://93860607&gt;

Reviewed by Brent Fulgham.

This patch addresses the regression reported in rdar://96083869. The
regression was an indirect consequnce of the initial patch.
WebCore::placeChildInlineBoxesInBlockDirection and
LegacyInlineFlowBox::placeBoxesInBlockDirection both took in a maxAscent
value as part of their arguments which had a type of int. However, the
actualy argument that was being passed into these methods was of type
LayoutUnit. Before the patch there were no issues since the logical
height and widths were being rounded to ints. Once the patch was
introduced this was no longer true and an incorrect type coercion was
occurring causing the incorrect value to be used in computations within
these methods. By changing the type to LayoutUnit we completely avoid
this issue and use the correct values.

Below is the description from the initial patch.

When computing the logical height/width for a replaced element, we may
need to calculate it using (used width) / (intrinsic ratio) or
(used height) / (intrinsic ratio). In either case, these are the
used values according to the spec, which means they may be floats.

Currently, WebKit performs these computations, but it also rounds them
leading to an incorrect value. By removing these calls to round, the
computed value should line up with what is specified in the spec.

As a result of this change, several tests needed to be rebaselined since
the generated content was slightly different. In the test cases, there
were certain elements that computed a slightly different width/height
value, which resulted in different dimensions that were used in the
rendering as well as in the output of the tests. Depending on how the
dimensions of an element changed, this could have in turn caused a
nearby element to get rendered slightly differetly as well. For example,
if an element had a slight increase in its height, this could have
caused a neighboring element to have a slightly different position in
the y dimension. The more of these interactions/differences there are
would mean we would see a greater difference in the expected vs actual
output (e.g. css1/text_properties/vertical_align.html).

However, this change should be seen as a progression when comparing the
tests with the results of other browsers. By loading up the tests in
other browsers, you can see that the dimensions that are calculated
are more in line with the new values generated with this patch.

Referenced spec: <a href="https://www.w3.org/TR/CSS21/visudet.html#propdef-height">https://www.w3.org/TR/CSS21/visudet.html#propdef-height</a>

* Source/WebCore/rendering/LegacyInlineFlowBox.cpp:
(WebCore::placeChildInlineBoxesInBlockDirection):
(WebCore::LegacyInlineFlowBox::placeBoxesInBlockDirection):
* Source/WebCore/rendering/LegacyInlineFlowBox.h:
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::resolveWidthForRatio):
(WebCore::RenderReplaced::computeReplacedLogicalWidth const):
(WebCore::resolveHeightForRatio):
(WebCore::RenderReplaced::computeReplacedLogicalHeight const):
* LayoutTests/fast/block/positioning/replaced-inside-fixed-top-bottom-expected.txt:
* LayoutTests/fast/writing-mode/block-level-images-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/replaced-fractional-height-from-aspect-ratio-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/replaced-elements/attributes-for-embedded-content-and-images/img-aspect-ratio-expected.txt:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/css2.1/t090501-c414-flt-03-b-g-expected.txt:
* LayoutTests/platform/glib/fast/block/positioning/replaced-inside-fixed-top-bottom-expected.txt:
* LayoutTests/platform/glib/tables/mozilla_expected_failures/bugs/bug85016-expected.txt:
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/gtk/css1/box_properties/width-expected.txt:
* LayoutTests/platform/gtk/css1/formatting_model/replaced_elements-expected.txt:
* LayoutTests/platform/gtk/css1/text_properties/vertical_align-expected.txt:
* LayoutTests/platform/gtk/css2.1/t090501-c414-flt-03-b-g-expected.png:
* LayoutTests/platform/gtk/fast/repaint/block-layout-inline-children-replaced-expected.png:
* LayoutTests/platform/gtk/fast/repaint/block-layout-inline-children-replaced-expected.txt:
* LayoutTests/platform/gtk/fast/replaced/width100percent-image-expected.txt:
* LayoutTests/platform/gtk/tables/mozilla/bugs/bug14929-expected.txt:
* LayoutTests/platform/ios/css1/text_properties/vertical_align-expected.txt:
* LayoutTests/platform/ios/fast/replaced/width100percent-image-expected.txt:
* LayoutTests/platform/ios/tables/mozilla/bugs/bug14929-expected.txt:
* LayoutTests/platform/ios/tables/mozilla_expected_failures/bugs/bug85016-expected.txt:
* LayoutTests/platform/mac/css1/box_properties/width-expected.txt:
* LayoutTests/platform/mac/css1/formatting_model/replaced_elements-expected.txt:
* LayoutTests/platform/mac/css1/text_properties/vertical_align-expected.txt:
* LayoutTests/platform/mac/fast/repaint/block-layout-inline-children-replaced-expected.txt:
* LayoutTests/platform/mac/fast/replaced/width100percent-image-expected.txt:
* LayoutTests/platform/mac/svg/wicd/test-rightsizing-b-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug14929-expected.txt:
* LayoutTests/platform/mac/tables/mozilla_expected_failures/bugs/bug85016-expected.txt:
* LayoutTests/svg/css/max-width-2-expected.html:

* LayoutTests/fast/rendering/inline-flex-container-with-replaced-elements-expected.txt
* LayoutTests/fast/rendering/inline-flex-container-with-replaced-elements.html
* LayoutTests/fast/rendering/resources/
New test added from the initial regression report along with the images
that are used in the test.

Canonical link: <a href="https://commits.webkit.org/252583@main">https://commits.webkit.org/252583@main</a>
</pre>

brentfulgham reviewed (approved):
Looks good after correcting the "regression" (that was actually an existing bug exposed by your change).

sammygill committed:
getComputedStyle(img).height returns string of a rounded int not a float
https://bugs.webkit.org/show_bug.cgi?id=206161
<rdar://93860607>

Reviewed by Brent Fulgham.

This patch addresses the regression reported in rdar://96083869. The
regression was an indirect consequnce of the initial patch.
WebCore::placeChildInlineBoxesInBlockDirection and
LegacyInlineFlowBox::placeBoxesInBlockDirection both took in a maxAscent
value as part of their arguments which had a type of int. However, the
actualy argument that was being passed into these methods was of type
LayoutUnit. Before the patch there were no issues since the logical
height and widths were being rounded to ints. Once the patch was
introduced this was no longer true and an incorrect type coercion was
occurring causing the incorrect value to be used in computations within
these methods. By changing the type to LayoutUnit we completely avoid
this issue and use the correct values.

Below is the description from the initial patch.

When computing the logical height/width for a replaced element, we may
need to calculate it using (used width) / (intrinsic ratio) or
(used height) / (intrinsic ratio). In either case, these are the
used values according to the spec, which means they may be floats.

Currently, WebKit performs these computations, but it also rounds them
leading to an incorrect value. By removing these calls to round, the
computed value should line up with what is specified in the spec.

As a result of this change, several tests needed to be rebaselined since
the generated content was slightly different. In the test cases, there
were certain elements that computed a slightly different width/height
value, which resulted in different dimensions that were used in the
rendering as well as in the output of the tests. Depending on how the
dimensions of an element changed, this could have in turn caused a
nearby element to get rendered slightly differetly as well. For example,
if an element had a slight increase in its height, this could have
caused a neighboring element to have a slightly different position in
the y dimension. The more of these interactions/differences there are
would mean we would see a greater difference in the expected vs actual
output (e.g. css1/text_properties/vertical_align.html).

However, this change should be seen as a progression when comparing the
tests with the results of other browsers. By loading up the tests in
other browsers, you can see that the dimensions that are calculated
are more in line with the new values generated with this patch.

Referenced spec: https://www.w3.org/TR/CSS21/visudet.html#propdef-height

* Source/WebCore/rendering/LegacyInlineFlowBox.cpp:
(WebCore::placeChildInlineBoxesInBlockDirection):
(WebCore::LegacyInlineFlowBox::placeBoxesInBlockDirection):
* Source/WebCore/rendering/LegacyInlineFlowBox.h:
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::resolveWidthForRatio):
(WebCore::RenderReplaced::computeReplacedLogicalWidth const):
(WebCore::resolveHeightForRatio):
(WebCore::RenderReplaced::computeReplacedLogicalHeight const):
* LayoutTests/fast/block/positioning/replaced-inside-fixed-top-bottom-expected.txt:
* LayoutTests/fast/writing-mode/block-level-images-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/replaced-fractional-height-from-aspect-ratio-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/replaced-elements/attributes-for-embedded-content-and-images/img-aspect-ratio-expected.txt:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/css2.1/t090501-c414-flt-03-b-g-expected.txt:
* LayoutTests/platform/glib/fast/block/positioning/replaced-inside-fixed-top-bottom-expected.txt:
* LayoutTests/platform/glib/tables/mozilla_expected_failures/bugs/bug85016-expected.txt:
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/gtk/css1/box_properties/width-expected.txt:
* LayoutTests/platform/gtk/css1/formatting_model/replaced_elements-expected.txt:
* LayoutTests/platform/gtk/css1/text_properties/vertical_align-expected.txt:
* LayoutTests/platform/gtk/css2.1/t090501-c414-flt-03-b-g-expected.png:
* LayoutTests/platform/gtk/fast/repaint/block-layout-inline-children-replaced-expected.png:
* LayoutTests/platform/gtk/fast/repaint/block-layout-inline-children-replaced-expected.txt:
* LayoutTests/platform/gtk/fast/replaced/width100percent-image-expected.txt:
* LayoutTests/platform/gtk/tables/mozilla/bugs/bug14929-expected.txt:
* LayoutTests/platform/ios/css1/text_properties/vertical_align-expected.txt:
* LayoutTests/platform/ios/fast/replaced/width100percent-image-expected.txt:
* LayoutTests/platform/ios/tables/mozilla/bugs/bug14929-expected.txt:
* LayoutTests/platform/ios/tables/mozilla_expected_failures/bugs/bug85016-expected.txt:
* LayoutTests/platform/mac/css1/box_properties/width-expected.txt:
* LayoutTests/platform/mac/css1/formatting_model/replaced_elements-expected.txt:
* LayoutTests/platform/mac/css1/text_properties/vertical_align-expected.txt:
* LayoutTests/platform/mac/fast/repaint/block-layout-inline-children-replaced-expected.txt:
* LayoutTests/platform/mac/fast/replaced/width100percent-image-expected.txt:
* LayoutTests/platform/mac/svg/wicd/test-rightsizing-b-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug14929-expected.txt:
* LayoutTests/platform/mac/tables/mozilla_expected_failures/bugs/bug85016-expected.txt:
* LayoutTests/svg/css/max-width-2-expected.html:

* LayoutTests/fast/rendering/inline-flex-container-with-replaced-elements-expected.txt
* LayoutTests/fast/rendering/inline-flex-container-with-replaced-elements.html
* LayoutTests/fast/rendering/resources/
New test added from the initial regression report along with the images
that are used in the test.

Canonical link: https://commits.webkit.org/252583@main
